biogeophys/FatesPlantHydraulicsMod.F90

```diff
@@ -4557,448 +4596,418 @@ subroutine MatSolve2D(site_hydr,cohort,cohort_hydr, &
+             !    searching.
+             ! 3) the residual is not 0, and either
+             !    we have taken too many newton steps or the solver won't return
+             !    a real solution.
+             !    Shorten time-step, reset time to 0, reset relaxation factors
+             !    and try a new round of newton (if not exceeded)
 
-            write(fates_log(),'(10x,a)') '---  Convergence Failure  ---'
-            write(fates_log(),'(4x,a,1pe11.4,2(a,i6),1pe11.4)') 'Equation Maximum Residual = ', &
-                 residual_amax,' Node = ',nsd, 'pft = ',ft, 'qtop: ',qtop
 
-            ! If we have not exceeded our max number
-            ! of retrying rounds of Newton iterations, reduce
-            ! time and try a new round
-            if( nsteps < max_newton_rounds ) then
+             if( icnv == icnv_fail_round ) then
+                 
+                 ! If the newton iteration fails, we go back
+                 ! to restart the time-stepping loop with shorter sub-steps.
+                 ! Therefore, we set the time elapsed (tm) to zero,
+                 ! shorten the timstep (dtime) and re-initialize the water
+                 ! contents to the starting amount.
+
+                 if(reset_on_fail) then
+                     tm  = 0._r8
+                     th_node(:)         = th_node_init(:)
+                     th_node_prev(:)    = th_node_init(:)
+                     cohort_hydr%iterh1 = 0
+                 else
+                     tm                 = tm - dtime
+                     th_node(:)         = th_node_prev(:)
+                     !* No need to update the th_node_prev, it is the
+                     !  same since we are just re-starting the current
+                     !  step
+                 end if
+                 nsteps             = nsteps + 1
+                 dtime              = dtime * dtime_rf
+                 rlfx_plnt0         = rlfx_plnt_init*rlfx_plnt_shrink**real(nsteps,r8)
+                 rlfx_soil0         = rlfx_soil_init*rlfx_soil_shrink**real(nsteps,r8)
+                 rlfx_plnt          = rlfx_plnt0
+                 rlfx_soil          = rlfx_soil0 
+                 nwtn_iter          = 0
+                 cohort_hydr%iterh1 = cohort_hydr%iterh1 + 1
+                 cycle outerloop
 
-               tm = tm - dtime
-               nsteps = nsteps + 1
+             else
 
-               write(fates_log(),*) 'fates hydraulics, MatSolve2D:'
-               write(fates_log(),'(4x,a,1pe11.4,1x,2a,1pe11.4,1x,a)')       &
-                    'Time Step Reduced From ',dtime,'s',' to ', &
-                    min(dtime * dtime_rf,tmx-tm),'s'
-               
-               dtime = min(dtime * dtime_rf,tmx-tm)
-               
-               do k = 1,site_hydr%num_nodes
-                  th_node(k) = th_node_init(k)
-               enddo
+                 ! On the last iteration, we temporarily lower the bar (if opted for)
```

Comment:
not an expert on the 2-D numerical solver but I assume they are robust given that you already did the test

Reply:
Good question.  How robust, is robust enough to release?  I guess that depends on how much testing we have done, and what the expected use case is.  The testing on the 2D solver has been more elaborate than many pull requests, in that we have done site level evaluation over many years... however, I personally  have not looked at many different sites (just bci).  Maybe @pnlfang has looked at other sites by now.
My take is that the 2d solver is an experimental feature right now, and should not be used by users who are not specifically interested in hypothesis testing or comparing the 2d solve itself.  Normal model users should use the 1D until we have a better sense of how stable the 2d is.  
The 2d solver is off by default, and is not switched on in the parameter file or the namelist, its a parameter constant logical here: https://github.com/NGEET/fates/blob/master/main/FatesHydraulicsMemMod.F90#L13  
This should protect users from trying this feature if they do not realize its experimental nature yet.

Reply:
Good points and it definitely provide opportunities to explore further for different sites in the future.